sklearn/utils/validation.py

```diff
@@ -1018,6 +1019,151 @@ def check_scalar(x, name, target_type, min_val=None, max_val=None):
         raise ValueError('`{}`= {}, must be <= {}.'.format(name, x, max_val))
 
 
+def _check_psd_eigenvalues(lambdas, warn_on_zeros=False):
```

Comment:
This warns under several circumstances, I'd probably rather call the parameter `warn` or something.

Reply:
I think @smarie had a reason for that, probably related to his follow-up PR https://github.com/scikit-learn/scikit-learn/pull/12069?

Reply:
Yes, but it has gone beyond simply checking for small values.

Reply:
let me have a look

Reply:
`_check_psd_eigenvalues` automatically sets to a proper zero all eigenvalues that are extremely small. I need this feature in PR #12069 , because it relies on an approximate method that very often stops at very small numbers instead of proper zeros.

But then I thought that in some (most?) cases, the caller might be interested in knowing that this operation actually happened. So I added this argument, that is probably not well-named and should rather be renamed `bad_conditioning_warning`. Maybe its default value should be actually `True`, and PR #12069 should explicitly set it to false when using approximate methods.

What do you think ?

Reply:
I'd be happy with the renaming to `bad_conditioning_warning`.

If the estimators pass `False` here, would there be any way through the public API that the user sees these warnings?

Reply:
For the renaming: @NicolasHug do you agree too ?
For the default value to `True`: do you both agree ?

For your last question @adrinjalali : if an estimator chooses to pass `False`, it purposedly disables the warning because it thinks that having small non-zero eigenvalues is a normal behaviour (due to numerical approximation). So no, there would not then be any way that the user sees this particular warning. The user would still see the other warnings though (significant negative eigenvalue).

Reply:
> For the renaming: @NicolasHug do you agree too ?

That means that parameter controls all the warnings, but I'm OK with it

>  For the default value to True: do you both agree ?

I'd leave it to False in this PR (better compatibility with master).
We can change it to True later in  #12069 if we need to.

THanks @smarie

Reply:
that means the user can pass a non-PSD matrix and the estimator would do something w/o giving any warnings?

This PR is supposed to fix an issue which says "raise Errors and Warnings according to eigenvalue decomposition numerical/conditioning issues", and we're not doing that if it's always False.

I'm happy with what this PR is doing in terms of fixing those issues in matrices, but we haven't figured how to show those messages to the user. I'd be more comfortable with it being always True for now, until people complain and we find a way to suppress them.

Reply:
> and we're not doing that if it's always False.

Yes, we are. We still raise a ValueError in the most problematic cases.

As you said, we're not sure what to do with those warnings. So having the default to False is the safe alternative. It's what master is currently doing. I don't think we should rely on users complaining about it to fix it.

Reply:
Ok, then we can merge this with the proposal you have, and open an issue immediately on finding a way to actually show the warnings which are never shown for now.

Reply:
sounds good to me

Reply:
I think that there was a misunderstanding here

Reply:
The `bad_conditioning_warning` option only applies to one warning (the last one) and does not at bypasses the other errors and warnings

Reply:
@adrinjalali 's point is that it should control all the warnings, not just one.

Also I think `bad_conditioning_warning` isn't a great name since we also raise errors when the conditionning is bad.

Reply:
I updated the PR in that sense, please have a look

Reply:
I re-updated the PR again with a very minor update: the comment in the kPCA estimator nos mentions `"make sure that the eigenvalues are ok and fix numerical issues"`.

Reply:
@smarie:

- the default should be false
- the parameter should control all the warnings
- there are  still references to `warn_on_zeros` left in the code and docs
- `bad_conditioning_warning` isn't a good name. Maybe `also_warn` would be better.

Reply:
Sorry @NicolasHug i skipped your comment, please find my answers here:

> it should control all the warnings, not just one.

There are only 2 kind of warnings in current proposal (the rest are errors) :

 - **significant negative eigenvalues**. This quite problematic for a PSD matrix so I think that the warning should always be displayed. 

 - **small positive eigenvalues**. This is not really a problem, , and it is very frequent in kPCA, since in kernel PCA I think that the matrix often has many trailing very small / zero eigenvalues (this is why the reconstruction using only a small number of eigenvectors is ok, hence the huge perf improvements of approximate decomposition methods). That's why it is really safe to ignore these warnings in kPCA estimator, as it is "nominal" for kPCA. Therefore we added the `bad_conditioning_warning` option to completely ignore the warning.

Now there is no warning currently for **small negative eigenvalues** (they are silently converted to zero without warning in all cases). Should we create one ? If so, should the `bad_conditioning_warning` parameter apply to both small positive and negative eigenvalues ? I think so, but I'd like to hear your thoughts.

Reply:
I'll leave you with @adrinjalali to discuss this. I have no strong opinion.

Reply:
ok. I'll wait for his reply and discuss with him tomorrow then. Thanks a lot for driving the discussions @NicolasHug !

Reply:
@adrinjalali : I updated the proposal according to the following thoughts:

 - since the term "bad conditioning" does not seem to really apply to matrices that have exact zero eigenvalues, I reverted the naming for something more explicit (`small_nonzeros_warning`).

 - I updated the code so that this warning is raised if there are small non-zero **negative or positive** eigenvalues (the warning was raised for positive small previously but not for negative small, which was not consistent)

Let me know if the result is consistent and intuitive enough so that we can mark this as resolved